### ISSUE TITLE
Custom reporters need the way to provide metadata for tests

### ISSUE THREAD

**Issue by vovsemenv** [vitest-dev/vitest#1951]
### Clear and concise description of the problem

In allure team want to create custom reporter for vitest but there is no way to provide metadata to each test such as linked issue on another important data.
Also we need the way to consume that metadata from reporter API.

### Suggested solution

for example playwright provide custom annotation API
```ts
test('user profile', async ({ page }) => {
  test.info().annotations.push({ type: 'issue', description: 'https://github.com/microsoft/playwright/issues/<some-issue>' });
  // ...
});
```

### Alternative

Maybe we can provide way to send  messages to reporter.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by sheremet-va** [comment]
Can you be more specific about usecase, and what exactly do you want? I do not know what `annotations` are.

**Comment by vovsemenv** [comment]
@sheremet-va Hello. We just need something to put metadata during tests and then consume this metadata in our custom reporter.
For example with this feature we can add rich markdown descriptions for each test or link Jira or GH tasks to specific tests.
This feature also will be useful for vitest own ui.

How i think it would look.
```ts
import { it, addMeta } from 'vitest'
it('should work', ({ foo }) => {
  addMeta("metaString")
  addMeta({type:"metaObj"})
})
```

```ts
export class CustomReporter implements Reporter {
  onFinished(files?: File[]) {
    console.log(files[0].tasks[0].meta) // ["metaString", {type:"metaObj"}]
}
```

**Comment by sheremet-va** [comment]
> How i think it would look.

We cannot do this with a global function, since we don't now what test is running with `test.concurrent`.

Why can't you use [context](https://vitest.dev/guide/test-context.html)?

```ts
it('some test', ({ context }) => {
  context.type = 'meta_obj'
})
```

We can technically add `setContext` function, if you prefer this style:
```ts
it('some test', ({ setContext }) => {
  setContext({ type: 'myObj' })
})
```

Right now context is not transferred to reporter, of course. But it can be implemented.

**Comment by vovsemenv** [comment]
> Why can't you use [context](https://vitest.dev/guide/test-context.html)?

I definitely can. I thought this solution was limited by the fact that context is not accessible in the reporter 😅 

>  But it can be implemented.

🔥🔥🔥

**Comment by sheremet-va** [comment]
Although I think it might be a breaking change, because you wouldn't be able to add a function on context (functions cannot be transferred between worked and main thread). So maybe indeed we need a new concept for this. 

I like the name `metadata`, how would you like to interact with it? Is it an array, as I see in your examples?

**Comment by vovsemenv** [comment]
> how would you like to interact with it

i like the way it realized in playwright. there you can do thing like
```ts
test("basic test", async (testArgs, testInfo) => {
  testInfo.attachments.push({
    name: "screenshot",
    body: await testArgs.page.screenshot(),
    contentType: "image/png",
  });

  testInfo.annotations.push({
    type: "issue-link",
    description: "https://github.com/vitest-dev/vitest/issues/1951",
  });
});
```

Wonder if we can do same thing like
```ts
it("basic test", async (context, testInfo) => {
  testInfo.metadata.push("something")
});
```

**Comment by vovsemenv** [comment]
`testInfo` in `TestFunction` will be really handy when we want to customize our test report from inside of the test
```
it("basic test", (context, testInfo) => {
  if(somCondition){
    testInfo.fail()
  }
});
```

This also allow us to access vite and vitest config from test. 
```ts
it("basic test", (context, testInfo) => {
  testInfo.viteConfg
  testInfo.vitestConfg
});
```

**Comment by sheremet-va** [comment]
> i like the way it implemented in playwright. there you can do thing like

We already provide `meta`:

```ts
test('some test', ({ meta }) => {
  console.log('metadata', meta)
})
```

> This also allow us to access vite and vitest config from test.

Vitest config is part of Vite config. It's also not fully possible, because it needs to be serializable, but some properties, like `plugins`, are not.

**Comment by vovsemenv** [comment]
> We already provide meta:

Is it not internal-only thing? It's ok to modify this meta from inside of test?
If so we really can provide something like meta.attachments to attach anything you want to this specific test.

**Comment by vovsemenv** [comment]
ofc we also need to expose the meta to reporter

**Comment by sheremet-va** [comment]
> > We already provide meta:
> 
> Is it not internal-only thing? It's ok to modify this meta from inside of test? If so we really can provide something like meta.attachments to attach anything you want to this specific test.

Well, it's not used internally. Vitest has direct access to it without needing context, but I guess the contents should be readonly (although it's not imposed).

**Comment by stefnotch** [comment]
We have a similar use-case for "sending test information to a custom reporter".

We're using Vitest for a university course, and all unit tests have a lot of helpful information for the students. We basically have an array of "steps", and then the reporter prints both the succeeded steps, and the (frequently customized) error message.

Currently, the only way of getting such information to the reporter is by doing horrible hacks. We've resorted to 
1. disabling the default reporter
2. abusing the error message and packing an entire object into it, as in `new Error(JSON.stringify({ steps, message }), { cause: e })`

![image](https://user-images.githubusercontent.com/10220080/235349914-4890b6a2-c378-403b-b2d2-b3a4e1b23b82.png)

In an ideal world, we could use the context to send over that information.

**Comment by sheremet-va** [comment]
We need to expose meta to the reporter. Currently only the "result" is transported to the reporter. We cannot reuse `context` for this because we cannot serialize it.

Also, it's only possible to pass down [clonable data](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm). I'm not sure if we can impose it with typescript.

**Comment by NKJe** [comment]
Is this resolved by https://github.com/vitest-dev/vitest/pull/3449?

**Comment by sheremet-va** [comment]
> Is this resolved by #3449?

Yes, you are right!

**Comment by vovsemenv** [comment]
Wow!!! That's awesome will check later
